### DISCUSSION TITLE
Might there be Java/Kotlin API (core or module) support in the future?

### DISCUSSION THREAD

**Discussion by minoism** [LSPosed/LSPlant#2] [Q&A] [answered]
很棒的项目！~~（也许这个应该放到讨论中，但这个仓库没开🤐）~~
我在自己的项目中使用 d8 来实现 java method hook，但最近有一些系统 runtime 方面的 hook 需求 d8 无法实现，因此发现了这个项目给我很大的惊喜，但目前似乎没有提供 java/kotlin 的 api? 请问有提供的打算吗？如果没有的话是否愿意接受 kotlin api 的贡献？（因为我发现几乎全部 lsp 项目都是用 java 编写的，kotlin 可能不受宠 😰）

> 我的项目内部使用着一套自己设计的不算太完善的 hook api，它是由 Kotlin 编译器和 IDE 实现的基于声明式和强类型理念的 api...大概
> 因此如果可以我很乐意补充这个项目缺少的易用 api . 如果有讨论区或者你们对此感兴趣，我可以提交一些设计细节，我可以随时开始 pr
>
> 它目前大概长这样
> ```kotlin
> class MainActivity : Suannai<Activity>(primaryHook = "com.example.main.MainActivity".type) {
>   // 反射名为 viewBinding 的字段
>   val Activity.viewBinding: ViewBinding by field()
>  
>   fun init() = hookConstructor(
>     head = {
>       require(arguments.isEmpty())
>       ...
>     },
>     tail = {
>       require(returns == Void)
>       ...
>     }
>   )
>   
>   // Hook and replace: 'public boolean isShowBottomNavigation(boolean show)'
>   fun changeBottomNavigationState(isShow: Boolean) = hook(
>     name = "isShowBottomNavigation",
>     body = {
>       // 替换传参
>       isShow = false
>       val activity = thisInstance<Activity>()
>       activity.viewBinding.apply {
>         if (activity.isFaild) { 
>           activity.recreate()
>           callOriginal(isShow)
>         }
>         bottomNavigationView.alpha = 0f
>         return(false)
>       }
>     }
>   )
>   
>   class ViewBinding : InstanceMock(actual = "com.example.databinding.ActivityMainBinding".type) {
>     val bottomNavigationView by field<View>(type = "com.example.views.ThemeBottomNavigationView")
>   }
> }
> ```

**Comment by yujincheng08** [comment, marked as answer]
因为需要用户提供 native hook 和 elf parser 所以无法提供纯 java/kt 接口，也因而直接放弃了提供 java/kt 接口了。不过我们的接口除了 Init 以外都能非常轻松就能封装到 Java/kt 上，所以其实没什么由库封装的必要？

**Comment by minoism** [comment]
原来是这样，实际上我对art hook的原理并不熟悉，刚开始看test时我还以为jni里面都是可以copy重用的样板代码

> 不过我们的接口除了 Init 以外都能非常轻松就能封装到 Java/kt 上，所以其实没什么由库封装的必要？

这么一说也有道理🥳，也许我应该将其作为一个额外的utils库而不是集成到core中

**Reply by vegedreamgagnoa** [reply]
@RinOrz let's build a library for this :) I have done this though.

**Reply by minoism** [reply]
@vegedreamgagnoa In fact, this repository work in progress, but I have not yet made it public

**Reply by Vendicated** [reply]
I wrote minimal Xposed java api for lsplant (only hooking implemented) based on Dobby. You can find it at https://github.com/Aliucord/hook if you would like to use it

**Reply by minoism** [reply]
> I wrote minimal Xposed java api for lsplant (only hooking implemented) based on Dobby. You can find it at [Aliucord/hook](https://github.com/Aliucord/hook) if you would like to use it

Nice work! But I don't know much about that license, I think I will take the time to look at it.
